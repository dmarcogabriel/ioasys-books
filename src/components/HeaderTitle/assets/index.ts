import LogoImage from './logo.png';
import LogoDarkImage from './logo-dark.png';

export {LogoDarkImage, LogoImage};
