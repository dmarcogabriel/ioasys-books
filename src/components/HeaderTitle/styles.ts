import styled from 'styled-components/native';
import {ThemeProps} from '../../types/theme.interface';

interface HeaderTitleContainerProps {
  noPadding: boolean;
  noMargin: boolean;
}

export const HeaderTitleContainer = styled.View`
  flex-direction: row;
  align-items: center;
  margin-bottom: ${({noMargin}: HeaderTitleContainerProps) =>
    noMargin ? 0 : '48px'};
  padding-top: ${({noPadding}: HeaderTitleContainerProps) =>
    noPadding ? 0 : '40px'};
`;

export const Logo = styled.Image``;

interface HeaderTextProps {
  color: 'light' | 'dark';
}

export const HeaderText = styled.Text`
  color: ${({theme, color}: ThemeProps & HeaderTextProps) =>
    theme.colors[color]};
  margin-left: 16px;
  font-size: 28px;
`;
