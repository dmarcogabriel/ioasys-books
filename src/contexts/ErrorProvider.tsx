import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
