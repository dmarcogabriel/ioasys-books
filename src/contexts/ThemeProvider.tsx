import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
