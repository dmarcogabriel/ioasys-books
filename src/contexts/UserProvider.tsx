import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
