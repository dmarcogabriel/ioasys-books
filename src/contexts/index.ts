export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
