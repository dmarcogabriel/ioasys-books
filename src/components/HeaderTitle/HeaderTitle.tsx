import React from 'react';
import {HeaderTitleContainer, Logo, HeaderText} from './styles';
import {LogoImage, LogoDarkImage} from './assets';

export interface HeaderTitleProps {
  color: 'light' | 'dark';
  noMargin?: boolean;
  noPadding?: boolean;
  testID?: string;
}

export const HeaderTitle = ({
  color = 'dark',
  noMargin = false,
  noPadding = false,
  testID = 'headerTitle',
}: HeaderTitleProps): JSX.Element => {
  return (
    <HeaderTitleContainer noMargin={noMargin} noPadding={noPadding}>
      <Logo source={color === 'dark' ? LogoDarkImage : LogoImage} />
      <HeaderText testID={testID} color={color}>
        Books
      </HeaderText>
    </HeaderTitleContainer>
  );
};
