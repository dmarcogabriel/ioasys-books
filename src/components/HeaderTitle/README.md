# Header Title

It's the application logo with two variations 'light' and 'dark'.

This component is built using `styled-components` 

## Usage

```js
import {HeaderTitle} from 'src/components';

...
  <HeaderTitle color="light" />
...
```

## API

| Prop         | Description                    | Type                           | Default       |
| ------------ | ------------------------------ | ------------------------------ | ------------- |
| **`color`**  | Color of the title content.    | Enum 'light' 'dark'          | 'light'       |
| **`testID`** | Id for unit testing            | string                         | 'headerTitle' |