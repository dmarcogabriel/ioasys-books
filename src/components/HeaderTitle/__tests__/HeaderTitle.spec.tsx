import React from 'react';
import {HeaderTitle, HeaderTitleProps} from '../HeaderTitle';
import {render} from '@testing-library/react-native';
import {ThemeProvider} from '../../../contexts';
import renderer from 'react-test-renderer';

const mockLogin = jest.fn((data: any) => data);
jest.mock('../../../hooks', () => ({
  useUser: () => ({
    login: mockLogin,
  }),
}));

const ComponentWrapper = (props: HeaderTitleProps) => (
  <ThemeProvider>
    <HeaderTitle {...props} />
  </ThemeProvider>
);

describe('components/HeaderTitle', () => {
  it('should pass on render header title dark', async () => {
    const {getByTestId} = render(<ComponentWrapper color="dark" />);

    const headerTitle = getByTestId('headerTitle');

    expect(headerTitle.children[0]).toBe('Books');
    expect(headerTitle).toHaveProp('color', 'dark');
  });

  it('should pass on render header title light', async () => {
    const {getByTestId} = render(<ComponentWrapper color="light" />);

    const headerTitle = getByTestId('headerTitle');

    expect(headerTitle.children[0]).toBe('Books');
    expect(headerTitle).toHaveProp('color', 'light');
  });

  it('should match snapshot', () => {
    expect(
      renderer.create(<ComponentWrapper color="light" />).toJSON(),
    ).toMatchSnapshot();
  });
});
