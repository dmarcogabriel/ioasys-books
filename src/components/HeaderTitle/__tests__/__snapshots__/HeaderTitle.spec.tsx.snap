// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/HeaderTitle should match snapshot 1`] = `
<View
  noMargin={false}
  noPadding={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 48,
        "paddingTop": 40,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/components/HeaderTitle/assets/logo.png",
      }
    }
    style={
      Array [
        Object {},
      ]
    }
  />
  <Text
    color="light"
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontSize": 28,
          "marginLeft": 16,
        },
      ]
    }
    testID="headerTitle"
  >
    Books
  </Text>
</View>
`;
